tests/robustness/watch.go

```diff
@@ -202,19 +203,31 @@ func validateOrdered(t *testing.T, report traffic.ClientReport) {
 	}
 }
 
-func validateUnique(t *testing.T, report traffic.ClientReport) {
-	type revisionKey struct {
-		revision int64
-		key      string
-	}
-	uniqueOperations := map[revisionKey]struct{}{}
-	for _, resp := range report.Watch {
-		for _, event := range resp.Events {
-			rk := revisionKey{key: event.Op.Key, revision: event.Revision}
-			if _, found := uniqueOperations[rk]; found {
-				t.Errorf("Broke watch guarantee: Unique - an event will never appear on a watch twice, key: %q, revision: %d, client: %d", rk.key, rk.revision, report.ClientId)
+func validateUnique(t *testing.T, expectUniqueRevision bool, report traffic.ClientReport) {
```

Comment:
There is a bit of duplication in this function that we could eliminate with something like below:

```go
func validateUnique(t *testing.T, expectUniqueRevision bool, report traffic.ClientReport) {
	uniqueOperations := map[interface{}]struct{}{}

	for _, resp := range report.Watch {
		for _, event := range resp.Events {
			var key interface{}
			if expectUniqueRevision {
				key = event.Revision
			} else {
				key = struct {
					revision int64
					key      string
				}{event.Revision, event.Op.Key}
			}

			if _, found := uniqueOperations[key]; found {
				t.Errorf("Broke watch guarantee: Unique - an event will never appear on a watch twice, key: %q, revision: %d, client: %d", event.Op.Key, event.Revision, report.ClientId)
			}
			uniqueOperations[key] = struct{}{}
		}
	}
}
```

Although there are some potential downsides for the interface type approach like reduced type safety and probably a little less performant.

This comment is just an idea, I'm happy for it to be discarded if it doesn't make sense in the context of etcd code priorities.

Reply:
Note: duplication is not bad and even good if it makes code simpler. Using `interface{}` could lead to code bloat due to casting and make it less readable. 

Still I like your example because usage `interface{}` is not bloating code too much.